### PR TITLE
Make shared IterationData a first class member of Workspace.

### PR DESCRIPTION
**Pull request by mzient**
Make operator traces thread-safe.

## Category:
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
This change does 2 things:
- it generalizes the mechanism of iteration data beyond operator traces
- it makes operator traces thread-safe

### Affected modules and functionalities:
Executor, Workspace, operator traces

### Key points relevant for the review:
N/A

### Tests:
- C API operator trace tests 
- Executor tests

- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/executor/executor_impl.cc [hunk 1]
```diff
@@ -157,6 +157,7 @@ void Executor<WorkspacePolicy, QueuePolicy>::SyncDevice() {
 
 template <typename WorkspacePolicy, typename QueuePolicy>
 void Executor<WorkspacePolicy, QueuePolicy>::RunCPUImpl(size_t iteration_id) {
+  GetCurrentIterationData(iteration_id).iter_data->iteration_index = iteration_id;
```
Comment (JanuszL) [inline dali/pipeline/executor/executor_impl.cc:160]: So it was not use by the CPU stage?

File: dali/pipeline/workspace/workspace.h [hunk 2]
```diff
@@ -603,16 +606,9 @@ class WorkspaceBase : public ArgumentWorkspace {
...
-
-  /**
-   * Get the operator trace map for all operators in the pipeline.
-   *
-   * @see operator_trace_map_t
-   */
-  DLL_PUBLIC const auto &GetOperatorTraceMap() const {
+  DLL_PUBLIC auto &GetOperatorTraces(const std::string &operator_name) const {
+    if (!operator_traces_)
+      operator_traces_ = &iter_data_->operator_traces.Get(operator_name);
```
Comment (JanuszL) [inline dali/pipeline/workspace/workspace.h:611]: So it is to avoid expensive locking?

File: dali/pipeline/workspace/workspace.h [hunk 3]
```diff
@@ -565,8 +565,7 @@ class WorkspaceBase : public ArgumentWorkspace {
    * @see operator_trace_map_t
    */
   DLL_PUBLIC void SetOperatorTrace(const std::string &trace_key, std::string trace_value) {
-    (*operator_traces_)[GetOperatorInstanceName()].insert_or_assign(
-            trace_key, std::move(trace_value));
+    GetOperatorTraces().insert_or_assign(trace_key, std::move(trace_value));
```
Comment (JanuszL) [inline dali/pipeline/workspace/workspace.h:568]: Should we mention here that this function is not thread safe as mentioned `OperatorTraces::Get`?
(other functions touching operator_trace_map_t in this file as well)

File: dali/pipeline/executor/executor_impl.cc (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/executor/executor_impl.cc:160]: OK, it is a new thing.

File: dali/pipeline/workspace/workspace.h (hunk 3 above)
Comment (mzient) [reply dali/pipeline/workspace/workspace.h:568]: I don't think we assume that _anything_ in Workspace is thread-safe. Mentioning here that it isn't might create an illusion of thread-safety elsewhere.

File: dali/pipeline/workspace/workspace.h (hunk 2 above)
Comment (mzient) [reply dali/pipeline/workspace/workspace.h:611]: Yes, that's the point.

File: dali/pipeline/workspace/workspace.h (hunk 3 above)
Comment (JanuszL) [reply dali/pipeline/workspace/workspace.h:568]: Understood.

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**